src/langs/vala/Compiler.js

```diff
@@ -1,39 +1,33 @@
 import Gio from "gi://Gio";
 import GLib from "gi://GLib";
 import dbus_previewer from "../../Previewer/DBusPreviewer.js";
+import { decode } from "../../util.js";
 
-export default function Compiler(data_dir) {
-  const code_file = Gio.File.new_for_path(
-    GLib.build_filenamev([data_dir, "code.vala"]),
-  );
-  const module_file = Gio.File.new_for_path(
-    GLib.build_filenamev([data_dir, "libworkbenchcode.so"]),
-  );
+export default function Compiler({ session }) {
+  const { file } = session;
 
-  async function compile(code) {
+  const module_file = file.get_child("libworkbenchcode.so");
+  const file_vala = file.get_child("main.vala");
+
+  async function compile() {
     let args;
+
     try {
+      const [contents] = await file_vala.load_contents_async(null);
+      const code = decode(contents);
       args = getValaCompilerArguments(code);
     } catch (error) {
       console.debug(error);
       return;
     }
-    await Promise.all([
-      code_file.replace_contents_async(
-        new GLib.Bytes(code || " "),
-        null,
-        false,
-        Gio.FileCreateFlags.NONE,
-        null,
-      ),
```

Comment:
I'm not sure why we were duplicating the existing vala code, but I'm not passing directly the buffer file (`sessions/x/main.vala`) to the compiler.

Makes sense?

Reply:
I think it was a problem when the file contains nothing, so "". That was prevented with replacing it with a " ".